src/shared/ui/pickerPrompter.ts

```diff
@@ -0,0 +1,255 @@
+
+/** Additional options to configure the `QuickPick` beyond the standard API */
+interface AdditionalQuickPickOptions<T = never> {
+    title?: string
+    value?: string
+    step?: number
+    placeholder?: string
+    totalSteps?: number
+    buttons?: PrompterButtons<T>
+    customInputSettings?: {
+        label: string
+        transform: (v: string) => PromptResult<T>
+    }
+}
+
+export type ExtendedQuickPickOptions<T> = Omit<vscode.QuickPickOptions, 'buttons' | 'canPickMany' | 'placeHolder'> &
+    AdditionalQuickPickOptions<T>
+
+export const DEFAULT_QUICKPICK_OPTIONS: vscode.QuickPickOptions = {
+    ignoreFocusOut: true,
+}
+
+type QuickPickData<T> = PromptResult<T> | (() => Promise<PromptResult<T>>)
+type LabelQuickPickItem<T> = vscode.QuickPickItem & { label: T }
+
+/**
+ * Attaches additional information as `data` to a QuickPickItem. Alternatively, `data` can be a function that
+ * returns a Promise, evaluated after the user selects the item.
+ */
+export type DataQuickPickItem<T> = vscode.QuickPickItem & { data: QuickPickData<T> }
+export type DataQuickPick<T> = Omit<vscode.QuickPick<DataQuickPickItem<T>>, 'buttons'> & { buttons: PrompterButtons<T> }
+
+const CUSTOM_USER_INPUT = Symbol()
+
+function isDataQuickPickItem(obj: any): obj is DataQuickPickItem<any> {
+    return typeof obj === 'object' && typeof (obj as vscode.QuickPickItem).label === 'string' && 'data' in obj
+}
+
+/**
+ * Creates a new QuickPick using special DataQuickPickItem interfaces. Information that should be returned when
+ * the user selects an item should be placed in the `data` property of each item. If only the `label` is desired,
+ * use `createLabelQuickPick` instead.
+ *
+ * @param items An array or a Promise for items.
+ * @param options Customizes the QuickPick and QuickPickPrompter.
+ * @returns A QuickPickPrompter. This can be used directly with the `prompt` method or can be fed into a Wizard.
+ */
+export function createQuickPick<T>(
+    items: DataQuickPickItem<T>[] | Promise<DataQuickPickItem<T>[] | undefined>,
+    options?: ExtendedQuickPickOptions<T>
+): QuickPickPrompter<T> {
+    const picker = vscode.window.createQuickPick<DataQuickPickItem<T>>() as DataQuickPick<T>
+    options = { ...DEFAULT_QUICKPICK_OPTIONS, ...options }
+    applyPrimitives(picker, { ...DEFAULT_QUICKPICK_OPTIONS, ...options })
+
+    const prompter =
+        options.customInputSettings !== undefined
+            ? new CustomQuickPickPrompter<T>(
+                  picker,
+                  options.customInputSettings.label,
+                  options.customInputSettings.transform
+              )
+            : new QuickPickPrompter<T>(picker)
+
+    if (items instanceof Promise) {
+        makePickerAysnc(picker, items)
+    } else {
+        picker.items = items
+    }
+
+    return prompter
+}
+
+/** Creates a QuickPick from normal QuickPickItems, using the `label` as the return value. */
+export function createLabelQuickPick<T extends string>(
```

Comment:
Why are we extending string?

Reply:
The generic type `extends` keyword is _not_ an object extension. This specifies that the generic type has at _least_ the same base structure as the `string` type.

Reply:
So the only place this is brought into concern is https://github.com/aws/aws-toolkit-vscode/pull/1874/files/72b1f7e2b7631e4310ec3abfe4bbd2246580e7b7#diff-f3213813c66557d1271643b441552dac6c1eddae5283a62839f8851416464fafR35 , correct? I struggle to think of why you'd want a label with additional properties when you're already defining a datatype.

Reply:
This is strictly a convenience function that creates the QuickPickItems in the desired shape to be consumed by the `QuickPickPrompter`. So instead of writing:
```js
createQuickPick([
{ label: 'Option1', data: 'Option1' },
{ label: 'Option2', data: 'Option2' }
])
```
You may write:
```js
createLabelQuickPick([{ label: 'Option1' }, { label: 'Option2' }])
```
I originally wanted `createQuickPick` to handle both cases automatically, but this made the types overly complicated .

Reply:
I understand the point of this function, I just don't understand any case where `label` will be `string & ...`. Dump the extended type, remove the T from `LabelQuickPickItem`, and just deal with strings everywhere.

Reply:
Oh duh good catch! I didn't even realize `label` was being redefined...
This blunder was from when I was trying to have the types automatically infer whether the items were of shape:
```js
{ label: string, data?: T }
```
I must've mistakenly placed `T` in label and just assumed everything was good.

Reply:
Errr wait. I remember now why I had the generic type `T`. This is for when there exists a type that is a union of string literals. The generic type is there to infer correctly that the literal types are a string. Though I'm not sure if the ergonomics of this are worth the added type complexity.